python/paddle/tensor/einsum.py

```diff
@@ -664,7 +666,139 @@ def plan_einsum(operands, g_view, g_shape, g_supports, g_count, n_bcast):
     return plan
 
 
+def preprocess(equation, *operands):
+    """
+    check equation / raise error, default right labels generation
+    """
+    equation = equation.replace(" ", "")
+    nop = len(operands)
+    assert nop > 0, "At least one operand is expected."
+
+    # Part the equation to left hand side and right hand side
+    lhs, *rhs = equation.lower().replace(' ', '').split('->')
+    assert len(rhs) < 2, "Invalid equation: multiple `->` were found."
+
+    labels = parse_labels(lhs, operands)
+    # Note, we distinguish between 'ij->' and 'ij' by setting rhs to '' and None
+    rhs = rhs[0] if rhs else None
+    if rhs is None:
+        rhs = rhs_inference(lhs, rhs)
+
+    assert len(lhs.split(',')) == len(operands), (
+        f"Invalid equation: the number of operands is {len(operands)}, "
+        f"but found {len(lhs.split(','))} segments in the label equation.")
+
+    if '...' in lhs and '...' not in rhs:
+        assert False, f'Invalid equation: missing ellipsis in output labels.'
+
+    if len(list(filter(has_duplicated_labels, lhs.split(',')))) > 0:
+        assert False, f'Duplicate labels are not supported.'
+
+    if has_duplicated_labels(rhs) > 0:
+        assert False, f'Invalid equation: duplicate output labels are found.'
+
+    return lhs, rhs, labels
+
+
+def parse_fake_shape(equation, operands, labels):
+    """ 
+    this shape is just used for operands planning. may differ with the original shape.
+    for example: 
+    ... is replaced by 1
+    -1  is replaced by 1
+    Results
+    -------
+    list of shape
+    """
+    import collections
+    shaped = collections.namedtuple('shaped', ['shape'])
+
+    def fake_shape(label, op):
+        assert len(op.shape) == len(
+            label), "length of shape and length of label must be the same."
+        fakes = [s for i, (l, s) in enumerate(zip(label, op.shape)) if l != '.']
+        fakes = list(map(abs, fakes))  # make -1 -> 1
+        if '.' in label:
+            fakes.insert(label.index('.'), 1)
+        return shaped(fakes)
+
+    out = list(map(fake_shape, labels, operands))
+    return out
+
+
+def rhs_inference(lhs, rhs):
```

Comment:
```suggestion
def rhs_inference(lhs):
```
这个函数似乎不需要 rhs 作为参数，函数代码里也应该移除 rhs is None的判断？